### PR TITLE
Merge from master to features

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
new pr

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -187,6 +210,9 @@
 .. _@Vogtinator: https://github.com/Vogtinator
 .. _@blueyed: https://github.com/blueyed
 .. _@fengxx: https://github.com/fengxx
+.. _@bagerard: https://github.com/bagerard
+.. _@davehunt: https://github.com/davehunt
```
Comment (The-Compiler) [inline CHANGELOG.rst]:
```
INFO CHANGELOG.rst:242 Duplicate explicit target name: "@davehunt".
```

(he's already on line 216/242)

**Comment by nicoddemus** [comment]
Hmm it failed on Windows:

```
_________________ ERROR collecting testing/python/metafunc.py _________________
_pytest\python.py:374: in fget
    return self._obj
   AttributeError: 'Module' object has no attribute '_obj'
During handling of the above exception, another exception occurred:
...
.tox\py35-xdist\lib\site-packages\hypothesis\internal\charmap.py:77: in categories
    cm = charmap()
.tox\py35-xdist\lib\site-packages\hypothesis\internal\charmap.py:63: in charmap
    os.rename(tmpfile, f)
FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'C:\\projects\\pytest\\.hypothesis\\tmp\\tmp9lzkh68w' -> 'C:\\projects\\pytest\\.hypothesis\\unicodedata\\8.0.0\\charmap.pickle.gz'
_________________ ERROR collecting testing/python/metafunc.py _________________
```

The first error (`'Module' object has no attribute '_obj'`) might be related to this [comment](https://github.com/pytest-dev/pytest/pull/1673#issuecomment-230647568), although I find the error about `hypothisis` suspicious, specially because it happend on an xdist environment...

**Comment by RonnyPfannschmidt** [comment]
at second glance it seems like a hypothesis bug @DRMacIver might know more

**Comment by flub** [comment]
So can we merge this anyway?  But every single PR will show this issue?  How wise is it to make all PRs dependent on an external project?

**Comment by The-Compiler** [comment]
To be honest I'd rather wait a day or two to see what @drmaciver has to say about it - and if it really turns out to be a hypothesis issue and we don't find a quick fix, just blacklist the hypothesis tests for now, merge this, and then investigate afterwards.

> How wise is it to make all PRs dependent on an external project?

Do you mean in general, or if we'd merge this?

**Comment by nicoddemus** [comment]
> To be honest I'd rather wait a day or two to see what @drmaciver has to say about it - and if it really turns out to be a hypothesis issue and we don't find a quick fix, just blacklist the hypothesis tests for now, merge this, and then investigate afterwards.

This merge is long overdue now, so I would rather do it now rather than wait a few more days. I propose we report the issue to hypothesis repository and `xfail` the test which depends on Hypothesis, but only when running on `xdist`:

```python
xdist = request.config.pluginmanager.getplugin('xdist')
if xdist:
    pytest.xfail(reason='bad interaction with xdist, waiting on HypothesisWorks/hypothesis-python/#X ')
```

> How wise is it to make all PRs dependent on an external project?

All PRs already depend on a bunch of external projects: `nose`, `mock` or even `colorama`, just to name a few. Hypothesis is no different IMO. Bugs in external projects are part of life. :grin:

**Comment by flub** [comment]
@nicoddemus It is a collection error though, maybe we should be adding `.hypothesis` to `norecursedirs` in `tox.ini`?  It's more a stab in the dark then really understanding the issue tough.

Sure about the external dep, it's unavoidable I guess.  It's just a bit annoying that this seemingly unrelated thing is holding up what should be a routine merge.

**Comment by The-Compiler** [comment]
> It is a collection error though, maybe we should be adding `.hypothesis` to `norecursedirs` in `tox.ini`?  It's more a stab in the dark then really understanding the issue tough.

It happens at collection time because the `@hypothesis.given` decorator (which is in the stacktrace) gets run then.

**Comment by The-Compiler** [comment]
Oh! According to https://github.com/HypothesisWorks/hypothesis-python/issues/344 a fix is out in Hypothesis 3.4.1 (this is why I suggested waiting, @DRMacIver is incredibly quick with fixing bugs :laughing:).

Since we don't seem to pin test dependency versions, rerunning the build now.

**Comment by nicoddemus** [comment]
> @DRMacIver is incredibly quick with fixing bugs

Oh I didn't know that, otherwise I might have suggested the same! :grin: 

Thanks @DRMacIver!

**Comment by nicoddemus** [comment]
I'm hitting that "Merge" button the second I see that AppVeyor build green! :laughing:

**Comment by nicoddemus** [comment]
Thanks everyone! :+1:

**Comment by DRMacIver** [comment]
> It happens at collection time because the @hypothesis.given decorator (which is in the stacktrace) gets run then.

Hmm. It's interesting that it errors there though. It's *supposed* to defer as much as possible to test execution time to speed up file loading and localize errors better.

**Pull request merged**